### ISSUE TITLE
Syntax error caused by autocorrection

### ISSUE THREAD

**Issue by r7kamura** [rubocop/rubocop#11445]
When I try to autocorrect the code below, I get a syntax error.

```ruby
# frozen_string_literal: true

a[*b { c; d }] # comment
```

--------

## Expected behavior

At least, I would like to see the code remain valid and not result in a syntax error.

## Actual behavior

A syntax error occurs during the autocorrection process.

```
$ rubocop --autocorrect --debug example.rb
For /tmp/rubocop1: configuration from /home/r7kamura/.rubocop.yml
Default configuration from /home/r7kamura/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.43.0/config/default.yml
Use parallel by default.
Skipping parallel inspection: only a single file needs inspection
Inspecting 1 file
Scanning /tmp/rubocop1/example.rb
Loading cache from /home/r7kamura/.cache/rubocop_cache/4130c4cbc1c979342a39b3f0b2a1634d2a9b9751/6d7a3b621ca1730e04accd938619e4bdab66cfb1/30fb28d7e7140b6f7342c2cfa8b582b7b23b2a2d
E

Offenses:

example.rb:3:6: C: [Corrected] Style/BlockDelimiters: Avoid using {...} for multi-line blocks.
a[*b { c
     ^
example.rb:3:7: C: [Corrected] Layout/TrailingWhitespace: Trailing whitespace detected.
a[*b { 
      ^
example.rb:3:8: C: [Corrected] Layout/MultilineBlockLayout: Block body expression is on the same line as the block start.
a[*b { c ...
       ^
example.rb:3:9: C: [Corrected] Style/Semicolon: Do not use semicolons to terminate expressions.
a[*b { c; d }] # comment
        ^
example.rb:4:1: C: [Corrected] Layout/IndentationWidth: Use 2 (not 5) spaces for indentation.
     c
^^^^^
example.rb:4:2: C: [Corrected] Layout/IndentationConsistency: Inconsistent indentation detected.
 d }] # comment
 ^
example.rb:4:4: C: [Corrected] Layout/BlockEndNewline: Expression at 4, 4 should be on its own line.
 d }] # comment
   ^
example.rb:5:8: C: [Corrected] Layout/IndentationConsistency: Inconsistent indentation detected.
       d
       ^
example.rb:8:1: E: Lint/Syntax: unexpected token $end
(Using Ruby 2.6 parser; configure using TargetRubyVersion parameter, under AllCops)

1 file inspected, 9 offenses detected, 8 offenses corrected
Finished in 0.10505970008671284 seconds
```

and the following code will be left:

```ruby
# frozen_string_literal: true

a[*# comment
b do
  c
     d
end
```

At some point `]` is lost.

## Steps to reproduce the problem

You can reproduce this by preparing the above code and running `rubocop -a`.

## RuboCop version

```
$ rubocop -V
1.43.0 (using Parser 3.2.0.0, rubocop-ast 1.24.1, running on ruby 3.1.2) [x86_64-linux]
```
